src/typing/flow_js.ml

```diff
@@ -4236,6 +4239,18 @@ let rec __flow cx ((l: Type.t), (u: Type.use_t)) trace =
       let new_obj = ObjT (reason, {objtype with flags}) in
       rec_flow_t cx trace (new_obj, t)
 
+    (*********************************************************)
+    (* objects can be marked as read-only using $ReadOnly<T> *)
+    (*********************************************************)
+
+    | (ObjT (reason_o, objtype), UseT(_, ReadOnlyT (reason_op, t))) ->
+      let desc = RReadOnly (desc_of_reason reason_o) in
+      let reason = replace_reason_const desc reason_op in
+
+      let flags = {frozen = true; sealed = Sealed; exact = true;} in
+      let new_obj = ObjT (reason, {objtype with flags}) in
```

Comment:
You probably do also want to mark the all the [fields](https://github.com/facebook/flow/blob/master/src/typing/type.ml#L638) covariant here to reap the benefits of read-only-ness

Should also decide now: If you have an ObjT that inherits a `proto_t`, should this operation re-generate the entire proto chain as read-only as well? I suspect this behavior could get surprising if it did not freeze every accessible field...